subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/AbstractModuleComponentResolveMetadata.java

```diff
@@ -188,6 +193,14 @@ public ModuleComponentArtifactMetadata artifact(String type, @Nullable String ex
         return platformOwners;
     }
 
+    @Override
+    public VariantDerivationStrategy getVariantDerivationStrategy() {
+        if (variantDerivationStrategy == null) {
```

Comment:
I do not understand this getter. It seems linked to the `variantDerivationStrategy = null;` but I am pretty sure we could always figure out a value or always use the value from `getVariantMetadataRules().getVariantDerivationStrategy()`.
In short I think we should make this simpler.

Reply:
This is again because of the conversion to mutable metadata, which does _not_ keep the strategy. I chose not to leak the strategy into the mutable version of component metadata (the immutable version is already enough, I think). Alternatively we need to store the derivation strategy in the mutable version too. That's a larger change that can be avoided by this strategy.

Reply:
I ended up adding it to mutable metadata too, for consistency.